keychain.h

```diff
@@ -0,0 +1,70 @@
+#include <assert.h>
+
+#include <optional>
+#include <string>
+#include <variant>
+
+namespace keychain {
+
+class Keychain {
+   public:
+    struct Error {
+        std::string message;
+    };
+    using Success = std::monostate;
+    using Password = std::string;
+    using Result = std::variant<Success, Password, Error>;
+
+    static Result getPassword(const std::string &service,
+                              const std::string &user);
+    static Result setPassword(const std::string &service,
+                              const std::string &user,
+                              const std::string &password);
+    static Result deletePassword(const std::string &service,
+                                 const std::string &user);
+
+   public:
+    /*! \brief Construct a Keychain specific to a service/user combination
+     *
+     * \param service Identifier for the service whose password should be stored
+     *                or updated. Note that a platform specific service name
+     *                will be used based on the provided identifier.
+     *                \sa getServiceName
+     * \param user
+     */
+    Keychain(const std::string &service, const std::string &user);
+
+    std::optional<std::string> getPassword();
+    bool setPassword(const std::string &password);
+    bool deletePassword();
+
+    //! Return the error resulting from the most recent call to getPassword,
+    //! addPassword, or deletePassword, if any
+    std::optional<std::string> lastError();
+
+   private:
+    //! \brief Create a platform specific service identifier
+    std::string makeServiceName(const std::string &service);
+
+    void clearError();
+
+    template <typename F>
+    bool updatePassword(F &&updateFun) {
+        clearError();
+
+        Result result = updateFun();
+
+        // should either return a Success or an Error
+        assert(!std::holds_alternative<Keychain::Success>(result));
+
+        auto err = std::get_if<Keychain::Error>(&result);
+        if (err) _lastError = err->message;
+        return !err;
+    }
```

Comment:
I'm not at all in love with this, but I don't want to duplicate the code in set and delete password. I considered `bool updatePassword(const std::optional<Password>&)`, but that means that updatePassword needs to reverse-engineer which static function to call. This way, `setPassword` and `deletePassword` need to know their namesake.

Reply:
You might want to move this into the cpp. I don’t see a need to have this exposed to the public, even though I see that it simplifies things...

Reply:
That would be great, but it's templated :(

Reply:
> That would be great, but it's templated :(

Who cares? It's a private method... Ah! because of `_lastError`. Mmh... `std::function<>` could fix this at some runtime cost. By the way: I recently learnt that `std::function<>` is a prime example for "type erasure", that's why it allows you to move (an otherwise templated) object to the cpp...

Reply:
But, I believe, you could also do:

```C++
class Keychain {
  private:
    template<typename Fun>
    bool updatePassword(Fun f);
};
```

```C++
// in cpp
template<typename Fun>
Keychain::updatePassword<Fun>(Fun f) {
//...
}
```